php/libraries/NDB_BVL_Instrument.class.inc

```diff
@@ -1331,15 +1331,15 @@ class NDB_BVL_Instrument extends NDB_Page
     /**
      * Run a XIN rule on an individual element
      *
-     * @param string $controller The element which affects the validity of this
+     * @param ?string $controller The element which affects the validity of this
```

Comment:
I'm confused about how this works with "NEVER_REQUIRED" as the PR describes. Wouldn't this change simply make instruments always claim that the rule is broken due to lines 1355-1356?

Reply:
@ridz1208 @driusan my interpretation of this change is at line 1303 will result in the $ElementResult having the value of false because the operator case would be ==. Otherwise without this PR the function XINRunRuleFunction doesn't except a $controller being null. Am I missing something?

Reply:
But doesn't it having a value of false mean it'll be flagged as being required when it's explicitly flagged as "NEVER_REQUIRED"? (The whole logic of this function is confusing.)

Reply:
@ridz1208 Could you clarify the summary of changes?

Reply:
@driusan it's confusing to me. I'm under the impression the The XINRunRuleFunction function comment states: `@return boolean true if element should be required, false otherwise.` and I'm currently thinking it does return false for a null $controller because I'm assuming $value isn't null because the function cannot take a null value with the type hinting. I'm also under the impression that the case is always '==' for NEVER_REQUIRED and never anything else to throw the LorisException.

Reply:
We can probably merge this since it just brings it (closer) to the state it was in before the type hinting was added, but I think it would be a good idea to write some unit tests for this function.

Reply:
The way I understand it (and I barely do) is:

if a field is optional, to allow it to be empty we add a rule that saays `only require fieldX when fieldX==NEVER_REQUIRED` that is a statementthat males it always not required unless it is actually `NEVER_REQUIRED` in which case there is no error because it was filled in and it was required... fun stuff. There is no special case for the string `NEVER_REQUIRED` it could be `RIDA_IS_THE_BEST` and this would still work

So... yeah it works and without this change it will break for anyone that uses this logic to bypass xin rules